feat: Add no-api-reassignments rule (#33)
salesforce/eslint-plugin-lwc: docs/rules/no-api-reassignments.md

# Prevent public property reassignments (no-api-reassignments)

> Reassignment public properties from with the component is generally a bad practice. The public properties are should be set by the parent components and should never be mutated. This linting rule prevents component to reassignment the properties they expose.

Reviewer: ```suggestion
Public properties should be set by the parent component and never reassigned by the child. This linting rule prevents reassignment of a component's public properties.
```

## Rule details

Example of **incorrect** code:

```js
import { LightningElement, api } from 'lwc';

export default class Test extends LightningElement {
    @api foo;

    constructor() {
        super();
        this.foo = 1;
    }

    method() {
        this.foo = 1;

        return () => {
            this.foo = 1;
        };
    }
}
```

Example of **correct** code:

```js
export default class Test extends LightningElement {
    @api foo = 1;

    method() {
        return function() {

>             this.foo = 1;

Reviewer: I find this example very confusing.

Reviewer 2: method should probably be turned into a public getter, oterwise this can never be invoked

Author: Agreed, this example adds more confusion than anything else. I wanted to illustrate that `this.foo` is allowed is the function expression scope because we don't know what is the `this` value. I will remove it.

        };
    }
}
```

## When not to use it

> If the component need to reflect internal state changes via public property, this rule should not be used. A good example for this is a custom input component reflecting its internal value change to the `value` public property.

Reviewer 2: ```suggestion
If the component needs to reflect internal state changes via public property, this rule should not be used. A good example for this is a custom input component reflecting its internal value change to the `value` public property.
```

## Caveats

This rule suffers from the following caveats:

-   It expects that all the class methods are invoked with the component instance as the `this` value. If a class method is invoked with a different `this` and happens to reassign a property that is also declared as a public property, this rule reports a false positive.

> -   It doesn't track public property reassignment when the `this` value is aliased (eg. `const that = this`).

Reviewer: bind(), apply(), call(), etc are also not tracked, right?
